### PR TITLE
Unit tests to use a random port for the dashboard

### PR DESCRIPTION
**Pull request by crusaderky**
- Drastically reduce the amount of warning messages, complaining that port 8787 is occupied, at the end of the test logs
- Use gen_test / gen_cluster whenever possible
- Clean up most of the other warnings captured by pytest

### CODE REVIEW

File: distributed/cli/tests/test_dask_scheduler.py
Comment (crusaderky) [inline distributed/cli/tests/test_dask_scheduler.py:39]: 404 is tested by test_no_dashboard below

File: distributed/comm/tests/test_ucx.py
Comment (crusaderky) [inline distributed/comm/tests/test_ucx.py:4]: This whole module is broken at least against the latest version of ucp. It needs bugfixing and CI integration.

File: distributed/deploy/tests/test_local.py
Comment (crusaderky) [inline distributed/deploy/tests/test_local.py:978]: Drop backwards compatibility for dask < 2021.04.0

File: distributed/tests/test_client.py
Comment (crusaderky) [inline distributed/tests/test_client.py:1987]: Drop backwards compatibility for dask < 2021.04.0

File: distributed/comm/tests/test_ucx.py
Comment (mrocklin) [reply distributed/comm/tests/test_ucx.py:4]: cc @quasiben

**Comment by crusaderky** [comment]
CI output of a successful Ubuntu 3.7 run:
Master: 8186 rows
This PR: 2466 rows :sunglasses: 
cc @mrocklin @jrbourbeau

File: distributed/comm/tests/test_ucx.py
Comment (mrocklin) [reply distributed/comm/tests/test_ucx.py:4]: cc also @pentschev

Comment (pentschev) [reply distributed/comm/tests/test_ucx.py:4]: > This whole module is broken at least against the latest version of ucp. It needs bugfixing and CI integration.

Could you provide more details? How are you installing UCX-Py and UCX, and how are you launching the tests? What errors do you see?

We run those tests in https://github.com/rapidsai/ucx-py/blob/branch-0.21/ci/gpu/build.sh#L111 and they're currently passing [here](https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fucx-py%2Fprb%2Fucx-py-gpu-build/detail/ucx-py-gpu-build/3122/pipeline/) where we test both UCX 1.9 and master branch.

As for CI integration, this is in the scope of https://github.com/dask/community/issues/138 .

**Comment by fjetter** [comment]
You are removing a lot of pytest-asyncio markers. What's the reasoning here? Should we drop pytest-asyncio generally? If it is about the cleanup fixture, we could probably define it as an [autouse fixture](https://docs.pytest.org/en/6.2.x/fixture.html#autouse-fixtures-fixtures-you-don-t-have-to-request)

**Comment by crusaderky** [comment]
> You are removing a lot of pytest-asyncio markers. What's the reasoning here? Should we drop pytest-asyncio generally? If it is about the cleanup fixture, we could probably define it as an [autouse fixture](https://docs.pytest.org/en/6.2.x/fixture.html#autouse-fixtures-fixtures-you-don-t-have-to-request)

I'm replacing them with gen_test, which features both cleanup and timeout. Unlike the timeout of pytest-asyncio, gen_test cleanly aborts a single test and can produce a stack trace.

Auto-adding a cleanup fixture is a problem because (1) I suspect it could interact poorly with gen_cluster - to be tested - and (2) cleanup fails for some tests. While ideally it is desirable to have all tests decorated by cleanup,being forced to fix the underlying issue all the times may prove to be unpractical.

**Comment by fjetter** [comment]
My question mostly points to "should we abandon pytest-asyncio or do we see a future". That's an interesting question for future code reviews, even if we still have remnants in the code base.

Regarding timeout in pytest-asyncio, see also https://github.com/pytest-dev/pytest-asyncio/pull/216 which implements our gen_test timeout logic there directly

**Comment by crusaderky** [comment]
We should not use pytest-asyncio for as long as it doesn't implement a timeout, and when it does, we *could* consider cleaning up a bunch of in-house stuff from distributed.utils_test. There are challenges there; namely
1. have a blanket default timeout instead of having to specify it every time
2. asyncio event loop vs. tornado event loop
3. blanket cleanup
4. either have a way to selectively disable cleanup, or go through *all* cleanup issues of *all* tests (which may not be even possible)

The notable exception is parametrized tests and pytest fixtures. gen_cluster supports them, but gen_test doesn't (yet), so pytest-asyncio there remains the only alternative for the time being.

File: distributed/comm/tests/test_ucx.py
Comment (crusaderky) [reply distributed/comm/tests/test_ucx.py:4]: Apologies, I botched the packages installation. Should've RTFM.

Comment (pentschev) [reply distributed/comm/tests/test_ucx.py:4]: No worries, thanks for the update.

From the UCX side I can confirm these changes still work, so I'm approving on behalf of UCX changes only and leave other changes for those with better understanding than myself to judge.

**Review by pentschev — APPROVED** [review]

**Comment by crusaderky** [comment]
Ready for review and merge

**Review by jrbourbeau — COMMENTED** [review]
Thanks for the updates here @crusaderky. Generally they look good. I noticed there are seemingly more timeout-related test failures here than in other recent PRs. Is this something we should be concerned about?

**Comment by crusaderky** [comment]
@jrbourbeau of the 6 failures you're talking about, 3 were potentially caused by my PR, while the other 3 seems to be unrelated.
I've taken the occasion to revisit the slow markers and the timeouts for all the tests that take >15s on CI.

**Review by jacobtomlinson — COMMENTED** [review]
That test output looks great! So much less scrolling.

**Comment by mrocklin** [comment]
I think that we should consider reverting this after
https://github.com/dask/distributed/pull/4928 gets in

If I recall correctly we used to do random ports, but then we had lots of
hard to explain random failures that came about because a scheduler/worker
from a previous test managed to connect back to the scheduler/worker of a
more recent test.  This was hard to understand because of the randomness.
We then started to use the same ports everywhere so that this problem was
more obvious and this then led us to the various cleanup testing
infrastructure.

In general, my experience is that avoiding/working around problems in the
distributed CI works, but sometimes leads to hard to track down situations.
Randomness is generally something that I think we want to avoid.

**Comment by crusaderky** [comment]
@mrocklin I assume that with "revert" you're just referring to the instances to ``dashboard_address=":0"`` throughout the test modules. This PR does a wealth of cleanup on top of that.

**Comment by mrocklin** [comment]
Yes, sorry, that's all I meant.  I don't think that we should rely on
random ports to hide any weirdness that our tests may generate.  We should
jump into and swim in that weirdness until it is fixed.  I'm hopeful that
the linked PR fixes that particular issue.

**Comment by crusaderky** [comment]
@jrbourbeau is there anything outstanding for this?

**Review by fjetter — APPROVED** [review]
FWIW I had a brief discussion with @jrbourbeau yesterday. My outstanding PR https://github.com/dask/distributed/pull/4928 is unfortunately still not ready to fix all port warning. It should not block this PR, especially not since there are other valuable cleanups. 

I'll merge once the CI jobs are through

**Pull request merged**
